File: src/main/java/com/sei/tamias/service/impl/FileTagServiceImpl.java
package com.sei.tamias.service.impl;

import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import com.sei.tamias.db.entity.FileInfo;
import com.sei.tamias.db.entity.FileTag;
import com.sei.tamias.db.entity.FileTagRelation;
import com.sei.tamias.db.mapper.FileTagMapper;
import com.sei.tamias.service.IFileTagRelationService;
import com.sei.tamias.service.IFileTagService;
import kotlin.Pair;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import javax.annotation.Resource;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import static com.sei.tamias.util.Options.*;
import static java.util.stream.Collectors.*;
/**
 * @author : gaoxiaodong04
 * @program : tamias
 * @date : 2020-11-09 17:05
 * @description :
 */
@Service
public class FileTagServiceImpl extends ServiceImpl<FileTagMapper, FileTag> implements IFileTagService {

    @Resource
    private IFileTagRelationService relationService;

    @Override
    public List<FileTag> listByFileInfoId(@NotNull Long fileId) {
        // 首先找到文件标签关联表中fileId相等的行
        List<Long> tagIds = relationService.lambdaQuery()
                .ge(FileTagRelation::getFileId, fileId)
                .list()
                .stream()
                .map(FileTagRelation::getTagId)
                .collect(toList());
        // 根据tagIds批量传tag信息
        return tagIds.isEmpty()? emptyList() : listByIds(tagIds);
    }

    @Override
    public Map<Long, List<FileTag>> mapByFileInfoIds(@NotNull Collection<? extends Long> idList) {
        Map<Long, List<FileTagRelation>> tagMap = relationService.lambdaQuery()
                .select(FileTagRelation::getFileId, FileTagRelation::getTagId)
                .in(FileTagRelation::getFileId, idList)
                .list()
                .stream()
                .collect(groupingBy(FileTagRelation::getFileId));

        if(tagMap.isEmpty()) {
            return emptyMap();
        } else {
            return buildMap(m-> tagMap.forEach((k, v)->{
                List<Long> ids = v.stream().map(FileTagRelation::getTagId).collect(toList());
                List<FileTag> s = v.isEmpty()? emptyList() : listByIds(ids);
                m.put(k, s);
            }));
        }
    }

    @Override
    public List<FileTag> listByFileInfo(@NotNull FileInfo fileInfo) {
        return fileInfo.getId()==null? emptyList() : listByFileInfoId(fileInfo.getId());
    }

    @Override
    public Map<Long, List<FileTag>> mapByFileInfos(@NotNull Collection<? extends FileInfo> files) {
        List<Long> ids = files.stream().map(FileInfo::getId).collect(toList());
        return ids.isEmpty()? emptyMap() : mapByFileInfoIds(ids);
    }

    @Override
    public List<Pair<FileTag, Integer>> countByTags() {
        // todo:需要从关系表中搜索，根据tagId进行groupBy返回找到对应文件的个数
        return null;
    }

    @Override
    public Integer countByTag(@NotNull Long tagId) {
        return relationService.lambdaQuery()
                .eq(FileTagRelation::getTagId, tagId)
                .count();
    }
}

File: src/main/java/com/sei/tamias/db/service/impl/FileTagServiceImpl.kt
package com.sei.tamias.db.service.impl

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import com.sei.tamias.db.entity.FileInfo
import com.sei.tamias.db.entity.FileTag
import com.sei.tamias.db.entity.FileTagRelation
import com.sei.tamias.db.mapper.FileTagMapper
import com.sei.tamias.db.service.IFileTagRelationService
import com.sei.tamias.db.service.IFileTagService
import org.springframework.stereotype.Service
import javax.annotation.Resource

/**
 *
 *
 * 服务实现类
 *
 *
 * @author sei
 * @since 2020-11-06
 */
@Service
class FileTagServiceImpl : ServiceImpl<FileTagMapper, FileTag>(), IFileTagService {

    @Resource
    lateinit var relationService: IFileTagRelationService

    override fun listByFileInfoId(fileId: Long): List<FileTag> {
        // 首先找到文件标签关联表中fileId相等的行
        val tagIds = relationService.lambdaQuery()
                .ge(FileTagRelation::getFileId, fileId)
                .list()
                .map { it.tagId }
        // 根据tagIds批量传tag信息
        return if(tagIds.isEmpty()) emptyList() else listByIds(tagIds)
    }

    override fun mapByFileInfoIds(idList: Collection<Long>): Map<Long, List<FileTag>> {
        val tagMap: Map<Long, List<FileTagRelation>> = relationService.lambdaQuery()
                .select(FileTagRelation::getFileId, FileTagRelation::getTagId)
                .`in`(FileTagRelation::getFileId, idList)
                .list()
                .groupBy { it.fileId }
        return if(tagMap.isEmpty()) emptyMap() else {
            mutableMapOf<Long, List<FileTag>>().apply{
                tagMap.forEach { (k, v) ->
                    val s = if(v.isEmpty()) emptyList() else listByIds(v.map { it.tagId })
                    put(k, s)
                }
            }
        }
    }

    override fun listByFileInfo(fileInfo: FileInfo): List<FileTag> {
        return  if(fileInfo.id==null) emptyList() else listByFileInfoId(fileId = fileInfo.id)
    }

    override fun mapByFileInfos(files: Collection<FileInfo>): Map<Long, List<FileTag>> {
        val ids = files.map { it.id }
        return if(ids.isEmpty()) emptyMap() else mapByFileInfoIds(ids)
    }

    override fun countByTags(): List<Pair<FileTag, Int>> {
        TODO("Not yet implemented")

    }

    override fun countByTag(tagId: Long?): Int {
        return relationService.lambdaQuery()
                .eq(FileTagRelation::getTagId, tagId)
                .count()
    }

}